tests/utest_memory_usage.cpp

```diff
@@ -0,0 +1,267 @@
+ * accomplished under Cooperative Agreement Number FA8750-19-2-1000. The views
+ * and conclusions contained in this document are those of the authors and should
+ * not be interpreted as representing the official policies, either expressed or
+ * implied, of the United States Air Force or the U.S. Government. The U.S.
+ * Government is authorized to reproduce and distribute reprints for Government
+ * purposes notwithstanding any copyright notation herein.
+ * -------------------------------------------------------------------------- */
+#include <gtest/gtest.h>
+#include <spark_dsg/edge_attributes.h>
+#include <spark_dsg/edge_container.h>
+#include <spark_dsg/mesh.h>
+#include <spark_dsg/metadata.h>
+#include <spark_dsg/node_attributes.h>
+#include <spark_dsg/scene_graph.h>
+#include <spark_dsg/scene_graph_layer.h>
+#include <spark_dsg/scene_graph_node.h>
+#include <spark_dsg/serialization/graph_binary_serialization.h>
+
+// TODO(lschmid): These tests are a quick sanity check and might break on different
+// platforms or with some future changes.
+
+namespace spark_dsg::tests {
+
+TEST(MemoryUsage, Mesh) {
+  Mesh mesh;
+  size_t expected_size = sizeof(Mesh);  // 160 bytes
+  EXPECT_EQ(mesh.memoryUsage(), expected_size);
+
+  mesh.points.resize(1000);
+  expected_size += sizeof(Mesh::Pos) * 1000;  // 12160 bytes
+  EXPECT_EQ(mesh.memoryUsage(), expected_size);
+
+  mesh.colors.resize(1000);
+  expected_size += sizeof(Color) * 1000;  // 28160 bytes
+  EXPECT_EQ(mesh.memoryUsage(), expected_size);
+}
+
+TEST(MemoryUsage, Metadata) {
+  Metadata metadata;
+  size_t expected_size = sizeof(Metadata);  // 16 bytes.
+  EXPECT_EQ(metadata.memoryUsage(), expected_size);
+
+  nlohmann::json contents;
+  contents["key1"] = "value1";
+  contents["key2"] = 42;
+  contents["nested"] = {{"nkey1", "nvalue1"}, {"nkey2", 3.14}};
+  metadata.set(contents);
+
+  expected_size += contents.dump().size();  // 85 bytes.
+  EXPECT_EQ(metadata.memoryUsage(), expected_size);
+}
+
+TEST(MemoryUsage, EdgeAttributes) {
+  EdgeAttributes attrs;
+  size_t expected_size = sizeof(EdgeAttributes);  // 40 bytes.
+  EXPECT_EQ(attrs.memoryUsage(), expected_size);
+
+  nlohmann::json meta_contents;
+  meta_contents["description"] = "Test edge";
+  meta_contents["weight"] = 2.5;
+  attrs.metadata.set(meta_contents);
+
+  expected_size += meta_contents.dump().size();  // 80 bytes
+  EXPECT_EQ(attrs.memoryUsage(), expected_size);
+}
+
+TEST(MemoryUsage, EdgeContainer) {
+  EdgeContainer edge_container;
+  size_t expected_size = sizeof(EdgeContainer);  // 144 bytes
+  EXPECT_EQ(edge_container.memoryUsage(), expected_size);
+
+  // Add some edges
+  for (size_t i = 0; i < 100; ++i) {
+    edge_container.insert(i, i + 1, std::make_unique<EdgeAttributes>(i * 0.1));
+  }
+  expected_size += 100 * 100;
+  EXPECT_EQ(edge_container.memoryUsage(), expected_size);
+}
+
+TEST(MemoryUsage, NodeAttributes) {
+  NodeAttributes attrs;
+  size_t expected_size = 47;
```

Comment:
idk how to make this robust to the attributes changing, so this is probably fine, but worth thinking about in the future